\section{Introduction}

% How does an MSA model compare to BC model as the number and length of loci
% change?

% When we have multi-locus sequence data, should we analyze all the data under
% a biallelic-character model, or only keep (at most) one site per locus?

% How do MSA and BC models compare in term of robustness to data-acquistion
% biases?
% BC can be very sensitive \citep{Oaks2018ecoevolity}.
% Will MSA fair better?
% Information from each site under a BC model is the counts of alleles.
% Thus, it's not surprising these models are sensitive when these counts are off.
% BC gleans information from many sites to directly inform the species
% tree by integrating out the gene trees; we have very little information
% about any of the gene trees.
% In comparison, information from aligned sequences of linked sites can
% potentially contain much more information about the underlying gene tree.
% Thus, if an acquistion bias causes some tips to be missing, the
% information in the gene tree and branch lengths from the remaning
% sequences might allow us to recover.

Likelihood-based inference under the multispecies coalescent is a powerful 
framework for estimating species trees,  divergence times, and  effective 
population sizes in the presence of incomplete lineage sorting 
\citep{xuChallengesSpeciesTree2016}.
\jrocomment{More intro to phylogenetics/species trees here}

Likelihood-based approaches to species 
tree inference can be classified into two groups, based on how they model the 
evolution of orthologous DNA sites along gene trees within the species 
tree---those that assume (1) each site evolved along its own gene tree 
(i.e., each site is ``unlinked'') 
\citep{bryantInferringSpeciesTrees2012, maioPoMoAlleleFrequencyBased2015}, 
or (2) contiguous, linked sites evolved along a shared gene tree 
\citep{liuSpeciesTreesGene2007, ogilvieStarBEAST2BringsFaster2017, 
yangBPPProgramSpecies2015}. We will refer to these as unlinked and 
linked-character models, respectively. For both models, the gene tree of each 
locus (whether each locus is a single site or a segment of linked sites) 
is assumed to be independent of the gene 
trees of all other loci, conditional on the species tree.
Methods using linked character models become computationally expensive as the
number of loci grows large, due to the estimation or numerical integration of
all of the gene trees \citep{bryantInferringSpeciesTrees2012}.
Unlinked-character models on the other 
hand are more tractable for a large number of loci, because  estimating 
individual gene trees is avoided by integrating over all possible gene trees 
\citep{bryantInferringSpeciesTrees2012}.
Whereas unlinked-character models can accomodate a larger number of loci than
linked-character models, most genetic data sets comprise linked sites and
unlinked-character models are unable to utilize this information.

Reduced-representation genomic data sets acquired from high-throughput
instruments are becoming commonplace in phylogenetics \citep{Leache2017}, and
usually comprise hundreds to thousands of loci as short as 50 nucleotides long
and up to several thousand base pairs long.
Investigators are thus faced with decisions about how best to 
use their data to infer a species tree.
Should they use a linked-character method that assumes the sites within each
locus evolved along a shared gene tree?
Ideally, the answer would be ``yes,'' however this is not always
computationally feasible and the model could be violated by intralocus
recombination.
Alternatively, should investigators remove all but a single-nucleotide
polymorphism (SNP) from each locus and use an unlinked-character model?
Or, perhaps they should apply the unlinked-character method to all of their
sites, even if this violates the assumption that each site evolved along an
independent gene tree?
Little work has been done to help inform these types of decisions. 

An important consideration for choosing the character model to utilize are
sources of error and bias that result from reduced representation methods,
high throughput sequencing and the processing of these data. High throughput
sequencing is now almost invariably performed on Illumina sequencing platforms
which have been shown to have error rates as high as 0.25\% per base 
\citep{@pfeifferSystematicEvaluationError2018}. 
To avoid introducing this error into analyses performed with the data, it is not 
uncommon to filter out variants that are not found above some minimum frequency 
threshold \citep{rochetteStacksAnalyticalMethods2019, linckMinorAlleleFrequency2019}. 
This filtering can alter allele frequencies and introduce bias which has been 
shown to have an effect on estimates derived from these frequencies \citep{linckMinorAlleleFrequency2019}.

Sequence coverage---the mean number of unique reads representing each locus---is
often highly variable among loci and among individuals. Higher coverage increases 
confidence of calling variants and distinguishing variants from error. 
At low coverage true variants not present at frequencies meeting the minimum threshold
may be filtered by processing tools 

As linked-character models estimate the gene tree of each locus there is 
information contained in the gene trees to inform estimates of parameters in the model. 
Unlinked character models rely only on allele counts to inform parameter 
estimates. This difference could make parameter estimates under linked-character 
models more robust to error and bias in the sequence data.

% Singleton site patterns here refer to sites in the alignment where one gene 
% copy differs from all other gene copies in each species.


\section{Methods}
\subsection{Simulations}
We simulated data sets with three different locus lengths—1000, 500, and 250 
linked characters—and then sampled from these alignments to create alignments 
with  error. We simulated 100 replicate data sets for each locus length and 
subsequent sampling routine. All data sets contained a total of 100,000 
characters sampled from two diploid individuals (4 haploid gene copies) from each 
of two species. We chose this simple sample design to help ensure any 
differences in estimation accuracy or precision were due to the underlying 
models, and not due to differences in numerical algorithms for searching tree 
space.  We simulated two tip species trees under a pure birth process with a 
birth rate of 10 using the \python package
\dendropy
\citep[Version 4.40; dev branch commit eb69003;][]{Dendropy}.  
We drew population sizes for each branch of the species tree from a Gamma 
distribution with a shape of 5.0 and mean of 0.002. We simulated 100, 200, and 
400 gene trees for the 1000, 500, and 250 locus length data sets respectively 
using the contained coalescent implemented in \dendropy. In order to 
simulate data under a model compatible with the biallelic model implemented in 
\ecoevolity, we simulated alignments of linked biallelic character data. 
We simulated linked biallelic character alignments using
\seqgen (Version 1.3.4)
\citep{@rambautSeqGenApplicationMonte1997}
with a GTR model with base frequencies of A and C equal to 0 and base 
frequencies of G and T equal to 0.5. The transition rate for all base changes was 
0 with the exception of the rate between G and T which was 1.0. 

\subsection{Introducing Site-pattern Errors}
From each simulated dataset described above, we created four datasets by 
introducing two types of errors at two levels of severity. The first type of 
error we introduced was changing singleton character patterns (i.e., characters 
for which one gene copy was different from the gene copies) to invariant 
patterns by changing the singleton character state to match the other gene 
copies. We introduced this change with a probability of 0.2 and 0.4 to create 
two datasets from each simulated dataset. The second type of error we introduced 
was converting heterozygote haplotypes to homozygous. To do this, for each locus 
we randomly paired gene copies from within each species, and with a probability 
of 0.2 or 0.4 we randomly replaced one with the other. 

\subsection{Inference}
We performed Bayesian inference to estimate the divergence time between the two species and the effective 
population sizes of the root and tips of the species tree from each alignment 
under both an unlinked-character model and a linked-character model. For each 
model we parameterized the prior distributions on divergence time and effective
population size with the same parameters under which the data were generated. 
the same as the distributions generating the data.
We used \ecoevolity
\citep[Version 0.3.2; dev branch commit a7e9bf2;][]{Oaks2018ecoevolity}
to infer these parameters under the 
unlinked-character model.
% Say something about the changes to the way the root Ne is estimated?
We ran four independent chains of \ecoevolity with 75,000 steps and a sample 
frequency of 50 steps.
We used the \beast (Version 0.15.1) \citep{@ogilvieStarBEAST2BringsFaster2017} 
package in \beastcore (Version 2.5.2) \citep{@bouckaertBEASTSoftwarePlatform2014}
to infer these parameters under the linked-character model. 
We ran two independent chains of \beast with 20 million steps and a 
sample frequency of 5000 steps. 
To model sequence evolution in \beast in a comparable way to \ecoevolity we used 
the same GTR model under which the data were simulated as explained above.


\section{Results}

%Outline
% MCMC
% ESS and PSRF Trends
% Coverage
% Ecoevolity 1bp
% StarBEAST vs Ecoevolity Time
% StarBEAST vs Ecoevolity Ne


% MCMC
We conservatively discarded the first 501 samples from the posterior of each 
\ecoevolity chain leaving 1,000 samples for each of the four chains and a combined
total of 4000.
We conservatively discarded the first 201 samples from the posterior of each 
\beast chain leaving 3,800 samples for each of the two chains and a combined
total of 7,600.


% Coverage
The 95\% credible intervals for both divergence times and effective population 
sizes estimated from alignments without error in both \beast and \ecoevolity had 
the expected frequentist coverage in that the true value was within 95\% of the 
estimated credible intervals. This confirms that the sequence data were simulated
under the same model as those used for inference in \beast.


% ESS and PSRF
We computed the effective sample size (ESS) for each parameter estimate and considered
a value of 200 to indicate adequate mixing of a chain. 
We also computed the potential scale reduction factor (PSRF) for each estimate and following 
consider a value of less than 1.2 to indicate adequate 
convergence of independent chains \citep{gelman1998}. 
The effective sample size was greater than 200 for most \ecoevolity paramter estimates
regardless of locus length and type of or amount of error. 
Likewise the potential scale reduciton factor was less than 1.2 for most \ecoevolity parameter estimates.

The proportion of estimates with PSRF values less than 1.2 was very low across 
all \beast analyses indicating convergence of most chains.
The proportion of \beast divergence time estimates with ESS values less than 200 
increased at more recent divergence times as locus length decreased but were 
consistent across different types and degrees of error.
The proportion of \beast root effective populaation size estimates with ESS 
values less than 200 was high across all analyses. However it was low for most estimates
of effective population size at the tips of the tree. With estimates made with 250 
bp loci there was a high proportion of low ESS values for estimates of small 
effective population sizes at the tip of the tree.


% Time
For estimates of divergence time when there is no model violation and all data 
are used both \ecoevolity and \beast
have high accuracy and precision when all of the data are used. The accuracy and
precision of divergence time estimates with \ecoevolity remains high for most 
types and degrees of error except at very recent divergence times where accuracy
and precision decline with greater degrees of error.

When only a single SNP per locus is used, the accuracy and precision of \ecoevolity 
declines precipitously.

The accuracy and precision of \ecoevolity estimates using a single SNP per locus 
also declines as the number of loci decreases.        




% Effective Population Size





% Ecoevolity 1 bp







Time results for 1000bp loci (Figure~\ref{fig:time1000}).

Time results for 500bp loci (Figure~\ref{fig:time500}).

Time results for 250bp loci (Figure~\ref{fig:time250}).

\section{Discussion}

As expected the linked-sequence model was more robust to model violation.

But only in a small range of parameter space when divergence time is recent.

Ecoevolity performs better when all of the data are used as it results in a much larger 
amount of data. It has previously been shown that including all sites including 
constant sites has a great positive impact than simply addding more data that 
do not violate the model \citep{Oaks2018ecoevolity}

Unsurprisingly MCMC performance declines with decreasing decreasing locus length.
There is less information contained in the shorter loci to inform gene tree estimation 
and there would be expected to be more error in gene tree estimation.

Poor MCMC performance in \beast does not appear to correlate with poor parameter 
estimates. The distribution of estimates is generally as good or better than those 
from \ecoevolity. 


\section{Conclusions}

Practitioners may want to compare both models if the number of loci and taxa
permit it. Or maybe perform analyses on parts of the tree with recent divergence
time to compare results 
